plutus-core/plutus-ir/src/PlutusIR/Analysis/Dependencies.hs

```diff
@@ -140,13 +140,14 @@ bindingDeps b = case b of
         vDeps <- tyVarDeclDeps d
         tvDeps <- traverse tyVarDeclDeps tvs
         cstrDeps <- traverse varDeclDeps constrs
-        -- All the datatype bindings depend on each other since they can't be used separately. Consider
-        -- the identity function on a datatype type - it only uses the type variable, but the whole definition
-        -- will therefore be kept, and so we must consider any uses in e.g. the constructors as live.
+        -- Making a match depends on all constructors for the time being.
+        -- TODO: Erase impossible cases and their respective constructors.
```

Comment:
I've just got another run at this task and faced a new wall. Here is a summary:

Removing branches of unused constructors was very easy as expected (unless I'm an idiot who misses cases...). A straightforward solution is to add a `DeConstructorMap` to `DepState` in `PlutusIR.Analysis.Dependencies`:
``` haskell
type DeConstructorMap = Map.Map PLC.Unique [PLC.Unique]
```

Which maps a deconstructor's unique id to its constructors. We can then populate it with:
``` haskell
Map.insert (destr ^. PLC.theUnique) (fmap (view PLC.theUnique) constrs)
```

When processing the dependencies of `DatatypeBind`. We then convert from the mapping to `[PLC.Unique]` to the mapping to `[Bool]` in `PlutusIR.Transform.DeadCode`'s `calculateLiveness`, by replacing a constructor's unique by its liveness:
``` haskell
liveness = Set.fromList $ T.reachable Deps.Root depGraph
live u = Set.member (Deps.Variable u) liveness
destructors' = fmap (fmap live) destructors
```

We then promote the `MonadReader Ctx m` to a `MonadState` with also a list tracking the current constructor's liveness during the traversal of the tree:
``` haskell
type CurrentConstructors = [Bool]
```

Then in `processTerm` we can pattern match `Apply _ (Var _ n) _` to check if `n` here is a deconstructor; if it is, put a new state with its `[Bool]` constructors' liveness. Then during the upward walk for each `Apply _ f rhs` we check if this list is not empty, if yes check the head, if true (the constructor is alive), keep the whole term, if not (the constructor is dead) replace it with just `f` to effectively delete the `rhs`. We then pop the head and keep going. An empty list here means no destructor is being processed and the term can be kept as is.

`andPlc` is then compiled as wanted above, but then `traceComplex` showed up:
https://github.com/input-output-hk/plutus/blob/64b33d1b6f31a5807e259677f79adb82cc6213c8/plutus-tx-plugin/test/Plugin/Primitives/Spec.hs#L152-L153

It didn't type-check because there is no `Bool` in the whole program which deletes both branches! Being a curious pleb, I tried to delete this whole unused function instead of reporting first... It wasn't very challenging for this instance, I can simply add another list of dead `PLC.Unique` to the state, to fill with all data types that don't have a single living constructor.
``` haskell
type DeadUniques = [PLC.Unique]
```

Then in `processTerm` "unitify" all `LamAbs x _ (TyApp _ (TyVar _ n) _) _` with a dead `n`. Basically replacing the whole term with `x <$ Unit.unitval`. `traceComplex` is then nuked correctly, but that is only one of the many broken cases left. For a learning experience, I went all the way to hackily address the remaining cases, which include cases like this:
```
LamAbs x _ (TyApp _ (TyLam _ _ _ (TyApp _ (TyVar _ n) _)) _) _
```

Another notable case is when the dead type is used inside a `TermBind` that `processTerm` cannot touch. The hack is then to add `TermBind`s using a dead type to the `DeadUniques` list mentioned above. This includes cases like this:
```
TermBind _ strictness d@(VarDecl _ n (TyFun _ (TyApp _ (TyVar _ n') _) _)) rhs
```
Then in `processTerm` we can actually nuke dead `Var x n`!

In summary, I would be fired if I submitted a patch that pattern matches to pass tests like that. The correct solution is to just recursively walks down on a type like in `runTypeUsages`. Even then, the elephant in the room is that it's going to take too much effort to only improve standalone programs with no inputs. 

I think we'll have to go with an "incorrect" dependency to be useful for Cardano validator scripts regardless. Making deconstructors depend on constructors was very dumb (I now think that deconstructors don't need to depend on anything?) but there may still be useful rules. For example, if a program reduces to a function, all of its type arguments (and their type dependencies?) must stay intact, while other types in the program can still be truncated. Another ambitious effort might be truncating input types (like `ScriptContext`) to only keep fields that are used for the validator script. A node would then have to parse each script to truncate the input value correspondingly.

I think my main problem with all these ideas is that we cannot tell if they are actually useful for dApp developers to justify further efforts. Maybe I should try the other way around, to look at real-life scripts to see how we can shrink them down without breaking their semantics, then make gradual progress per case study. The near future demands are to support complex "raw" scripts and basic-but-non-trivial state machines.

@michaelpj What do you think?

Reply:
Again, I think there are several things going on here.

1. Making dependency analysis correct for datatypes. This is I think unambiguously good so long as we don't introduce more hacks.
2. I still think it would be better to do https://github.com/input-output-hk/plutus/issues/4147, which is much simpler and safer to implement.
3. We are in fact in a better situation than you think for scripts with arguments. See below.

I really suggest you do 1&2 rather than spending too much time on this, which continues to prove tricky and of questionable utility. 

> For example, if a program reduces to a function, all of its type arguments (and their type dependencies?) must stay intact, while other types in the program can still be truncated.

We could in principle do something like this and try and pick some set of types that are potentially used externally and so have to be kept intact. But this is a bit tricky, because the form of the script may not be totally obvious. We might have to ask the user to specify them.

> Maybe I should try the other way around, to look at real-life scripts to manually find dead code and optimization opportunities first, then make gradual progress per case study.

Yes, but this is not totally obvious. At least with https://github.com/input-output-hk/plutus/issues/4147 we have at least one known, clear usecase, which is good.

----

A validator
```
myValidator :: MyDatum -> MyRedeemer -> ScriptContext -> Bool
myValidator = ...
```
is actually wrapped up with conversions from `Data`, so in fact it's something like:
```
myValidator' :: Data -> Data -> Data -> ()
myValidator' d r c = myValidator (unsafeFromData d) (unsafeFromData r) (unsafeFromData c)
```
So we *do* have references to all the pieces of the types in question due to the `IsData` instances. It's only `Data` that looks unused - but it's a builtin type.

So I think we could get away with it, although it's still a bit questionable.

Reply:
> I really suggest you do 1&2 rather than spending too much time on this, which continues to prove tricky and of questionable utility.

This direction was pretty fun but yeah, too impractical for now. I'll continue with 1 & 2 as suggested. 

> We are in fact in a better situation than you think for scripts with arguments. See below.

I've got the same vibe given how much contract code stayed intact in the original stats I posted on this PR. At least we know this is not a dead-end with a potential revisit in the future. Just need to be more formal next time I guess.

> Yes, but this is not totally obvious. At least with #4147 we have at least one known, clear usecase, which is good.

I'll do both! We know many people both internally and externally who are coming up with creative ideas to reduce their real-life scripts as well. It should be a nice group effort.